### PR TITLE
add container name validation during creation and push

### LINKED ISSUE #3698: Odo fails when devfile component container name is really long
/kind bug

## What versions of software are you using?

**Operating System:**

**Output of `odo version`:** master

Kube has a resource name limit of 63 characters. Currently, we do not check if the devfile container component name is less than 63 characters before using it:
```
Maysuns-MacBook-Pro:nodejs-ex maysun$ odo push

Validation
 ✓  Validating the devfile [119014ns]

Creating Kubernetes resources for component madridnode
 ✗  Failed to start component with name madridnode. Error: Failed to create the component: unable to create or update component: unable to create Deployment madridnode: Deployment.apps "madridnode" is invalid: spec.template.spec.containers[0].name: Invalid value: "runtimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntime": must be no more than 63 characters
```

Opened a new issue because it requires some code refactoring that impacts some features that rely on the name match.

/area devfile

### PR DESCRIPTION
**Pull request by yangcao77**
Signed-off-by: Stephanie <yangcao@redhat.com>

**What type of PR is this?**
/kind bug
/area devfile

**What does does this PR do / why we need it**:
Add  validation for devfile container name during component creation, and push. Block the action if the container name is not a valid K8s resourece name. 

**Which issue(s) this PR fixes**:

Fixes #3698 

**PR acceptance criteria**:

- [ ] Unit test 

- [x] Integration test 

- [ ] Documentation 

- [ ] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:

test component creation:
create a devfile with a long container name, and run `odo create --devfile <devfile path>`
```
$ odo create --devfile ./devfile.yaml 
Validation
 ✓  Creating a devfile component from devfile path: /Users/stephanie/Documents/devfile-spring2/devfile.yaml [62590ns]
 ✓  Validating devfile component [277583ns]
 ✗  endpoint name "runtimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntime" is not valid, endpoint name should conform the following requirements: 
- Contain at most 63 characters
- Contain only lowercase alphanumeric characters or ‘-’
- Start with an alphanumeric character
- End with an alphanumeric character
- Must not contain all numeric values
```

test odo push:
`odo create nodejs --starter`
modify the devfile to use a long container name
run `odo push`
```
$ odo push
 ✗  Validating K8s Resources Name [132166ns]
 ✗  endpoint name "runtimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntime" is not valid, endpoint name should conform the following requirements: 
- Contain at most 63 characters
- Contain only lowercase alphanumeric characters or ‘-’
- Start with an alphanumeric character
- End with an alphanumeric character
- Must not contain all numeric values

```

### CODE REVIEW

**Comment by girishramnani** [comment]
@yangcao77 this seems like something that can be tested via unit test? So can we write a unit test instead of integration?

**Comment by yangcao77** [comment]
@girishramnani The validation code added is mainly inside the cli modules, so IMO integration test is the better place to test for. 
If we want to use unit test, I can only test the functionality of the util function `ValidateK8sResourceName`, not the actual cli behavior

File: pkg/odo/cli/component/push.go [hunk 1]
```diff
@@ -175,6 +176,28 @@ func (po *PushOptions) Validate() (err error) {
 	// If Devfile is present we do not need to validate the below S2I checks
 	// TODO: Perhaps one day move Devfile validation to here instead?
 	if util.CheckPathExists(po.DevfilePath) {
+		spinner := log.Spinner("Validating K8s Resources Name")
```
Comment (johnmcollier) [inline pkg/odo/cli/component/push.go:179]: What about just `Validating devfile component` as the spinner text, like `odo create`?

Comment (yangcao77) [reply pkg/odo/cli/component/push.go]: updated

File: pkg/odo/cli/component/push.go [hunk 2]
```diff
@@ -175,6 +176,28 @@ func (po *PushOptions) Validate() (err error) {
 	// If Devfile is present we do not need to validate the below S2I checks
 	// TODO: Perhaps one day move Devfile validation to here instead?
 	if util.CheckPathExists(po.DevfilePath) {
+		spinner := log.Spinner("Validating devfile component")
+		defer spinner.End(false)
+		err = util.ValidateK8sResourceName("component name", po.EnvSpecificInfo.GetName())
+		if err != nil {
+			return err
+		}
+		containerComponents := adaptersCommon.GetDevfileContainerComponents(po.Devfile.Data)
```
Comment (girishramnani) [inline pkg/odo/cli/component/push.go:185]: we can extract https://github.com/openshift/odo/pull/3959/files#diff-6e3a2602ffd394e9a73d4fbf1388be8dR185-R190 out into a helper function and use it in create as well.

Comment (girishramnani) [reply pkg/odo/cli/component/push.go:185]:
```
	containerComponents := adaptersCommon.GetDevfileContainerComponents(devObj.Data)
	for _, comp := range containerComponents {
		err := util.ValidateK8sResourceName("container name", comp.Name)
		if err != nil {
			return err
		}
	}
```

Comment (yangcao77) [reply pkg/odo/cli/component/push.go]: updated to use unit tests instead of integration tests

File: pkg/devfile/validate/validate.go [hunk 3]
```diff
@@ -44,3 +47,15 @@ func ValidateDevfileData(data interface{}) error {
 	return nil
 
 }
+
+// ValidateContatinerName validates whether the container name is valid for K8
+func ValidateContatinerName(devfileData data.DevfileData) error {
```
Comment (maysunfaisal) [inline pkg/devfile/validate/validate.go:52]: typo

Comment (yangcao77) [reply pkg/devfile/validate/validate.go]: updated

File: pkg/odo/cli/component/push.go [hunk 4]
```diff
@@ -175,6 +176,25 @@ func (po *PushOptions) Validate() (err error) {
 	// If Devfile is present we do not need to validate the below S2I checks
 	// TODO: Perhaps one day move Devfile validation to here instead?
 	if util.CheckPathExists(po.DevfilePath) {
+		spinner := log.Spinner("Validating devfile component")
+		defer spinner.End(false)
```
Comment (adisky) [inline pkg/odo/cli/component/push.go:180]: we should validate under one spinner
```
$ odo push
 ✓  Validating devfile component [91679ns]

Validation
 ✓  Validating the devfile [22104ns]
```

**Review by adisky — CHANGES_REQUESTED** [review]

File: pkg/odo/cli/component/push.go (hunk 4 above)
Comment (yangcao77) [reply pkg/odo/cli/component/push.go]: updated

**Comment by cdrage** [comment]
/lgtm

Reviewed and looks good! Thanks @yangcao77

**Pull request merged**